File: solutions/haskell/protein-translation/1/src/ProteinTranslation.hs
module ProteinTranslation(proteins) where

code :: String -> String
code "AUG" = "Methionine"
code "UUU" = "Phenylalanine"
code "UUC" = "Phenylalanine"
code "UUA" = "Leucine"
code "UUG" = "Leucine"
code "UCU" = "Serine"
code "UCC" = "Serine"
code "UCA" = "Serine"
code "UCG" = "Serine"
code "UAU" = "Tyrosine"
code "UAC" = "Tyrosine"
code "UGU" = "Cysteine"
code "UGC" = "Cysteine"
code "UGG" = "Tryptophan"
code "UAA" = "STOP"
code "UAG" = "STOP"
code "UGA" = "STOP"

proteins :: String -> Maybe [String]
proteins s
  | length s >= 3 = case code (take 3 s) of
     "STOP" -> Just []
     c      -> Just [c] <> proteins (drop 3 s)
  | s == "" = Just []      

File: solutions/erlang/protein-translation/1/src/protein_translation.erl
-module(protein_translation).

-export([proteins/1]).
-define(AMINO_ACIDS_CODONS, #{
  "AUG" => methionine,
  "UUU" => phenylalanine,
  "UUC" => phenylalanine,
  "UUA" => leucine,
  "UUG" => leucine,
  "UCU" => serine,
  "UCC" => serine,
  "UCA" => serine,
  "UCG" => serine,
  "UAU" => tyrosine,
  "UAC" => tyrosine,
  "UGG" => tryptophan,
  "UGU" => cysteine,
  "UGC" => cysteine,
  "UAA" => stop,
  "UAG" => stop,
  "UGA" => stop
}).

proteins([]) -> {ok, []};
proteins([A,B,C|T]) ->
  X = maps:find([A,B,C], ?AMINO_ACIDS_CODONS),
  case X of
    error ->      {error, badarg};
    {ok, stop} -> {ok, []}; 
    {ok, P} ->
      case proteins(T) of
	{ok, Ps} -> {ok, [P|Ps]};
        error -> {error, badarg};
        {error, badarg} -> {error, badarg}
      end
  end;
proteins([_|_]) -> {error, badarg}.

